File: python/lesson 14/trapping_rain_water.py
# https://leetcode.com/problems/trapping-rain-water/

class Solution:
    def trap(self, height: List[int]) -> int:
        leftMax = [0] * len(height)
        leftMax[0] = height[0]
        for i in range(1, len(height)):
            leftMax[i] = max([height[i], leftMax[i - 1]])

        rightMax = [0] * len(height)
        rightMax[len(height) - 1] = height[len(height) - 1]
        for i in range(len(height) - 2, 0, -1):
            rightMax[i] = max(height[i], rightMax[i + 1])
        
        result = 0
        for i in range(1, len(height) - 1):
            result += min(leftMax[i], rightMax[i]) - height[i]
        
        return result

File: ruby/lesson 14/trapping_rain_water.rb
# https://leetcode.com/problems/trapping-rain-water/

def trap(height)
    left_max = Array.new(height.length)
    left_max[0] = height[0]
    (1..height.length - 1).each { |i| left_max[i] = [height[i], left_max[i - 1]].max }

    right_max = Array.new(height.length)
    right_max[height.size - 1] = height[height.size - 1]    
    (height.length - 2).downto(0) { |i| right_max[i] = [height[i], right_max[i + 1]].max }
        
    result = 0
    (1..height.length - 1).each { |i| result = result + [left_max[i], right_max[i]].min - height[i] }
    result
end
